B4---C++-Programming/cpp_bomberman/annexes/doxygen/man/man3/XmlLoad.3
.TH "XmlLoad" 3 "Tue Jun 9 2015" "Version 0.53" "cpp_bomberman" \" -*- nroff -*-
.ad l
.nh
.SH NAME
XmlLoad \- load xml file  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <XmlLoad\&.hh>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBXmlLoad\fP (std::string const &filename)"
.br
.RI "\fI\fBXmlLoad\fP's constructor\&. \fP"
.ti -1c
.RI "\fB~XmlLoad\fP ()"
.br
.RI "\fI\fBXmlLoad\fP's destructor\&. \fP"
.ti -1c
.RI "void \fBloadSave\fP ()"
.br
.RI "\fISet \fBGameState\fP from xml file\&. \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "int \fBgetIntValue\fP (\fBtinyxml2::XMLElement\fP *root, std::string const &name) const "
.br
.RI "\fIget int from xml \fP"
.ti -1c
.RI "float \fBgetFloatValue\fP (\fBtinyxml2::XMLElement\fP *root, std::string const &name) const "
.br
.RI "\fIget float from xml \fP"
.ti -1c
.RI "unsigned int \fBgetUnValue\fP (\fBtinyxml2::XMLElement\fP *root, std::string const &name) const "
.br
.RI "\fIget unsigned int from xml \fP"
.ti -1c
.RI "std::vector< \fBACharacter\fP * > bombermans \fBloadBombers\fP (\fBtinyxml2::XMLElement\fP *root) const "
.br
.RI "\fIload bombermans from xml \fP"
.ti -1c
.RI "\fBPlayer\fP * \fBcreatePlayer\fP (\fBtinyxml2::XMLElement\fP *player) const "
.br
.RI "\fIload player from xml \fP"
.ti -1c
.RI "\fBIA\fP * \fBcreateAI\fP (\fBtinyxml2::XMLElement\fP *ai) const "
.br
.RI "\fIload AI from xml \fP"
.ti -1c
.RI "std::vector< \fBBomb\fP * > \fBcreateBombs\fP (\fBtinyxml2::XMLElement\fP *bombs) const "
.br
.RI "\fIload bombs from player/ai \fP"
.ti -1c
.RI "\fBMaze\fP * \fBcreateMaze\fP (\fBtinyxml2::XMLElement\fP *maze) const "
.br
.RI "\fIload maze from xml \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBtinyxml2::XMLDocument\fP \fBm_xmlDoc\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
load xml file 

Load a xml file and set \fBGameState\fP instance 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "XmlLoad::XmlLoad (std::string const & filename)"

.PP
\fBXmlLoad\fP's constructor\&. It will load our file and set a xml's object
.PP
\fBParameters:\fP
.RS 4
\fIfilename\fP xml file we want to load 
.RE
.PP

.SS "XmlLoad::~XmlLoad ()"

.PP
\fBXmlLoad\fP's destructor\&. It will close our file and delete our xml's object 
.SH "Member Function Documentation"
.PP 
.SS "\fBIA\fP* XmlLoad::createAI (\fBtinyxml2::XMLElement\fP * ai) const\fC [private]\fP"

.PP
load AI from xml It will create an AI from our xml AIs' list
.PP
\fBParameters:\fP
.RS 4
\fIai\fP pointer into a element of AIs' list 
.RE
.PP
\fBReturns:\fP
.RS 4
New player loaded/created from our xml 
.RE
.PP

.SS "std::vector<\fBBomb\fP *> XmlLoad::createBombs (\fBtinyxml2::XMLElement\fP * bombs) const\fC [private]\fP"

.PP
load bombs from player/ai It will create a bombs' vector from current player/ai
.PP
\fBParameters:\fP
.RS 4
\fIbombs\fP pointer into player/ai's bombs 
.RE
.PP
\fBReturns:\fP
.RS 4
bombs's vector loaded 
.RE
.PP

.SS "\fBMaze\fP* XmlLoad::createMaze (\fBtinyxml2::XMLElement\fP * maze) const\fC [private]\fP"

.PP
load maze from xml It will create the maze from our xml
.PP
\fBParameters:\fP
.RS 4
\fImaze\fP pointer into our xml 
.RE
.PP
\fBReturns:\fP
.RS 4
New maze loaded/created from our xml 
.RE
.PP

.SS "\fBPlayer\fP* XmlLoad::createPlayer (\fBtinyxml2::XMLElement\fP * player) const\fC [private]\fP"

.PP
load player from xml It will create a player from our xml players' list
.PP
\fBParameters:\fP
.RS 4
\fIplayer\fP pointer into a element of players' list 
.RE
.PP
\fBReturns:\fP
.RS 4
New player loaded/created from our xml 
.RE
.PP

.SS "float XmlLoad::getFloatValue (\fBtinyxml2::XMLElement\fP * root, std::string const & name) const\fC [private]\fP"

.PP
get float from xml It will search the first child element needed and return the value
.PP
\fBParameters:\fP
.RS 4
\fIroot\fP Current position into our xml file 
.br
\fIname\fP Name used in FirstChildElement to find our element 
.RE
.PP
\fBReturns:\fP
.RS 4
The value found in our FirstChildElement 
.RE
.PP

.SS "int XmlLoad::getIntValue (\fBtinyxml2::XMLElement\fP * root, std::string const & name) const\fC [private]\fP"

.PP
get int from xml It will search the first child element needed and return the value
.PP
\fBParameters:\fP
.RS 4
\fIroot\fP Current position into our xml file 
.br
\fIname\fP Name used in FirstChildElement to find our element 
.RE
.PP
\fBReturns:\fP
.RS 4
The value found in our FirstChildElement 
.RE
.PP

.SS "unsigned int XmlLoad::getUnValue (\fBtinyxml2::XMLElement\fP * root, std::string const & name) const\fC [private]\fP"

.PP
get unsigned int from xml It will search the first child element needed and return the value
.PP
\fBParameters:\fP
.RS 4
\fIroot\fP Current position into our xml file 
.br
\fIname\fP Name used in FirstChildElement to find our element 
.RE
.PP
\fBReturns:\fP
.RS 4
The value found in our FirstChildElement 
.RE
.PP

.SS "std::vector<\fBACharacter\fP *> bombermans XmlLoad::loadBombers (\fBtinyxml2::XMLElement\fP * root) const\fC [private]\fP"

.PP
load bombermans from xml It will load 'players' and 'AIs' elements\&. It will call createPlayer and createAI until the end
.PP
\fBParameters:\fP
.RS 4
\fIroot\fP the current pointer on 'save' tag (root tag in xml save) 
.RE
.PP
\fBReturns:\fP
.RS 4
Bombermans' vector with all players/AIs loaded 
.RE
.PP

.SS "void XmlLoad::loadSave ()"

.PP
Set \fBGameState\fP from xml file\&. Main function to read our xml file and set our \fBGameState\fP 
.SH "Member Data Documentation"
.PP 
.SS "\fBtinyxml2::XMLDocument\fP XmlLoad::m_xmlDoc\fC [private]\fP"
Xml object needed to load and read xml file 

.SH "Author"
.PP 
Generated automatically by Doxygen for cpp_bomberman from the source code\&.


B4---C++-Programming/cpp_bomberman/src/save/XmlLoad.cpp
#include "GameState.hh"
#include "XmlLoad.hh"
#include "Error.hpp"

XmlLoad::XmlLoad(std::string const &filename) {
  if(this->m_xmlDoc.LoadFile(filename.c_str()) != tinyxml2::XML_SUCCESS)
    throw ErrorBomberman("Unable to load save", "XMLLOAD");
}

XmlLoad::~XmlLoad() {
}

void XmlLoad::loadSave() {
  GameState::Instance()->setLoad(true);
  GameState::Instance()->setMaze(this->createMaze(this->m_xmlDoc.FirstChildElement("save")->FirstChildElement("maze")));
  GameState::Instance()->setPlayers(this->loadBombers(this->m_xmlDoc.FirstChildElement("save"), GameState::Instance()->getMaze().getWidth(), GameState::Instance()->getMaze().getHeight()));
  GameState::Instance()->setWidth(GameState::Instance()->getMaze().getWidth());
  GameState::Instance()->setHeight(GameState::Instance()->getMaze().getHeight());
}

std::vector<ACharacter *> XmlLoad::loadBombers(tinyxml2::XMLElement *root, int width, int height) const {
  std::vector<ACharacter *>bombermans;
  tinyxml2::XMLElement *players = root->FirstChildElement("players");
  tinyxml2::XMLElement *AIs = root->FirstChildElement("AIs");

  tinyxml2::XMLElement *pListPlayers = players->FirstChildElement("player");
  while (pListPlayers) {
    bombermans.push_back(this->createPlayer(pListPlayers));
    pListPlayers = pListPlayers->NextSiblingElement("player");
  }
  tinyxml2::XMLElement *pListAIs = AIs->FirstChildElement("AI");
  while (pListAIs) {
    bombermans.push_back(this->createAI(pListAIs, width, height));
    pListAIs = pListAIs->NextSiblingElement("AI");
  }
  return bombermans;
}

int XmlLoad::getIntValue(tinyxml2::XMLElement *root, std::string const &name) const {
  int value;

  root->FirstChildElement(name.c_str())->QueryIntText(&value);
  return value;
}

float XmlLoad::getFloatValue(tinyxml2::XMLElement *root, std::string const &name) const {
  float value;

  root->FirstChildElement(name.c_str())->QueryFloatText(&value);
  return value;
}

unsigned int XmlLoad::getUnIntValue(tinyxml2::XMLElement *root, std::string const &name) const {
  unsigned int value;

  root->FirstChildElement(name.c_str())->QueryUnsignedText(&value);
  return value;
}

bool XmlLoad::getBoolValue(tinyxml2::XMLElement *root, std::string const &name) const {
  bool value;

  root->QueryBoolAttribute(name.c_str(), &value);
  return value;
}

Player *XmlLoad::createPlayer(tinyxml2::XMLElement *player) const {
  Player        *newPlayer;

  newPlayer = new Player(std::string(player->FirstChildElement("name")->GetText()),
			 this->getUnIntValue(player, "position"));
  newPlayer->setMaxBombs(this->getIntValue(player, "maxBombs"));
  newPlayer->setState(this->getBoolValue(player, "state"));
  newPlayer->setDirection(static_cast<eDirection>(this->getIntValue(player, "direction")));
  newPlayer->setSpeed(this->getFloatValue(player, "speed"));
  newPlayer->setRangeBomb(this->getUnIntValue(player, "rangeBomb"));
  newPlayer->setTexturePath(std::string(player->FirstChildElement("texture")->GetText()));
  newPlayer->setScore(this->getUnIntValue(player, "score"));
  newPlayer->setKeyLeft(this->getIntValue(player->FirstChildElement("keys"), "left"));
  newPlayer->setKeyRight(this->getIntValue(player->FirstChildElement("keys"), "right"));
  newPlayer->setKeyUp(this->getIntValue(player->FirstChildElement("keys"), "up"));
  newPlayer->setKeyDown(this->getIntValue(player->FirstChildElement("keys"), "down"));
  newPlayer->setKeyBomb(this->getIntValue(player->FirstChildElement("keys"), "bomb"));
  newPlayer->setBombs(this->createBombs(player->FirstChildElement("bombs")));
  return (newPlayer);
}

IA  *XmlLoad::createAI(tinyxml2::XMLElement *ai, int width, int height) const {
  IA  *newAI;

  newAI = new IA(this->getUnIntValue(ai, "position"),
		 static_cast<eDifficulty>(this->getIntValue(ai, "difficulty")),
		 width, height);
  newAI->setMaxBombs(this->getIntValue(ai, "maxBombs"));
  newAI->setState(this->getBoolValue(ai, "state"));
  newAI->setDirection(static_cast<eDirection>(this->getIntValue(ai, "direction")));
  newAI->setSpeed(this->getFloatValue(ai, "speed"));
  newAI->setRangeBomb(this->getUnIntValue(ai, "rangeBomb"));
  newAI->setTexturePath(std::string(ai->FirstChildElement("texture")->GetText()));
  newAI->setBombs(this->createBombs(ai->FirstChildElement("bombs")));
  return (newAI);
}

std::list <Bomb *>  XmlLoad::createBombs(tinyxml2::XMLElement *bombs) const {
  std::list<Bomb *> newBombs;
  tinyxml2::XMLElement *pListBombs = bombs->FirstChildElement("bomb");
  while (pListBombs) {
    Bomb *newBomb = new Bomb(this->getUnIntValue(pListBombs, "position"));
    newBomb->setDelay(this->getUnIntValue(pListBombs, "time"));
    newBomb->setRange(this->getUnIntValue(pListBombs, "range"));
    newBomb->setTexturePath(std::string(pListBombs->FirstChildElement("texture")->GetText()));
    newBombs.push_back(newBomb);
    pListBombs = pListBombs->NextSiblingElement("bomb");
  }
  return newBombs;
}

Maze *XmlLoad::createMaze(tinyxml2::XMLElement *maze) const {
  Maze *newMaze;
  char *map;

  map = const_cast<char *>(maze->FirstChildElement("maze")->GetText());
  newMaze = new Maze(this->getIntValue(maze->FirstChildElement("size"), "width"),
  this->getIntValue(maze->FirstChildElement("size"), "height"), map);
  return newMaze;
}
